LG-11503 Getting started page for OIDC Guide rework (#381)
18F/identity-dev-docs: _pages/oidc/getting-started.md

---
title: OpenID Connect
lead: >
   <a class="usa-link usa-link usa-link--external" href="http://openid.net">OpenID Connect</a> is a simple identity layer built on top of the OAuth 2.0 protocol. Login.gov supports <a class="usa-link usa-link usa-link--external" href="http://openid.net/specs/openid-connect-core-1_0.html">version 1.0</a> of the specification and conforms to the <a class="usa-link usa-link usa-link--external" href="https://openid.net/wg/igov">iGov Profile</a>.
redirect_from:
  - /openid-connect/
sidenav:
  - text: Getting started
    href: "oidc/getting-started/"
    links:
      - text: Choosing an authentication method
        href: "#choosing-an-authentication-method"
      - text: Set up a Sandbox account
        href: "#set-up-a-sandbox-account"
      - text: Auto-discovery
        href: "#auto-discovery"
      - text: Example application
        href: "#example-application"
  - text: Authorization
    href: "oidc/authorization/"
  - text: Token
    href: "oidc/#token"
  - text: User info
    href: "oidc/#user-info"
  - text: Certificates
    href: "oidc/#certificates"
  - text: Logout
    href: "oidc/#logout"
  - text: Example application

>     href: "oidc/#example-application"

Reviewer: is there any way to reuse this sidenav between pages? having to change 2+ other pages if a heading gets renamed seems like a lot of work

Author: hmm not sure... each page nav will be slightly different because the page you are on will have the sub sections on the current page 'expanded' and sub sections on other pages hidden

---

> <div class="usa-alert usa-alert--warning">
> <div class="usa-alert__body">
> <p class="usa-alert__text">We do not support the OpenID Connect “implicit flow” with client_secret because it is not recommended by the OAuth group for security reasons</p>
> </div>
> </div>

Reviewer: We have the `alert` partial in this project so we ca replace this with
```suggestion
{% capture implicit_flow_warning %}
We do not support the OpenID Connect “implicit flow” with client_secret because it is not recommended by the OAuth group for security reasons
{% endcapture %}
{% include alert.html content=implicit_flow_warning alert_class="usa-alert--warning" %}
```

## Getting started

### Choosing an authentication method

Login.gov supports two ways of authenticating clients: **private_key_jwt** and **PKCE**.

- **private_key_jwt** (preferred for web apps)

>   The client sends a <a class="usa-link usa-link usa-link--external" href="https://jwt.io/">JSON Web Token</a>, or JWT, signed with a private key (minimum length of 2048 bits) when requesting access tokens. The corresponding public key is registered with the IdP ahead of time, similar to SAML.

Reviewer: ```suggestion
  The client sends a <a class="usa-link usa-link--external" href="https://jwt.io/">JSON Web Token</a>, or JWT, signed with a private key (minimum length of 2048 bits) when requesting access tokens. The corresponding public key is registered with the IdP ahead of time, similar to SAML.
```
`usa-link` was added here twice. Can we do these as plain markdown syntax links and then maybe modify our `content_typography.rb` to detect external links and add the right class there?

https://github.com/18F/identity-dev-docs/blob/6f5568c261ffacafb9cd535232aa9de8ca6a16cf/_plugins/content_typography.rb#L17

Author: sure that page is almost all markdown so I would rather keep it as markdown wherever possible

Author: On second thought I know how I would do this with Javascript but not sure how to deduce what the host is using just plain ruby. In JS I would just check `window.host`

Reviewer: We could see if the URL includes `site.baseurl` or not, if not, then it's probably external

Reviewer: Alternatively, you could do `[link title](https://example.com){:class="usa-link--external"}` syntax

Author: done!

Author: @zachmargolis  mind taking another look ?

    - `openssl req -nodes -x509 -days 365 -newkey rsa:2048 -keyout private.pem -out public.crt`
- Create an application, at which point you will need to decide between private_key_jwt or PKCE

It is important to note that your Login.gov production account and your Login.gov sandbox account are two separate accounts.

### Auto-discovery

Consistent with the specification, Login.gov provides a JSON endpoint for OpenID Connect auto-discovery at 
`/.well-known/openid-configuration`. 

Integration URI:
- [https://idp.int.identitysandbox.gov/.well-known/openid-configuration](https://idp.int.identitysandbox.gov/.well-known/openid-configuration)

Production URI:
- [https://secure.login.gov/.well-known/openid-configuration](https://secure.login.gov/.well-known/openid-configuration)

### Example Application

The Login.gov team has created an example client to speed up your development, all open source in the public domain: [identity-oidc-sinatra](https://github.com/18F/identity-oidc-sinatra){:class="usa-link--external"}.

> <a href="{{ site.baseurl }}/oidc/authorization/" class="usa-link margin-top-4">Next step: Authorization</a>

Reviewer: ```suggestion
[Next step: Authorization]({{ site.baseurl }}/oidc/authorization/){:class="margin-top-4"}
```
